### DISCUSSION TITLE
HX Option in NDI filters?

### DISCUSSION THREAD

**Discussion by jbcurler2010** [DistroAV/DistroAV#825] [General]
I realize there hasn't been much movement here on v5 but I am hopeful it has not been abandoned. As such, I am also hopeful that in the development process, there would be plans to add NDI HX as an option for NDI Output. Thanks.

**Comment by paulpv** [comment]
It is a licensing issue.
I have asked at:
https://forums.newtek.com/threads/ndi-hx-oss-requirements.167340/

**Comment by paulpv** [comment]
This is related to my RFC PR from 2022/05:
https://github.com/Palakis/obs-ndi/pull/738/

**Comment by MarximusMaximus** [comment]
What is the issue with the licensing? Do you have a direct link to the licenses in question?

**Reply by paulpv** [reply]
It is mentioned in my post on the newtek forums.

obs-ndi is a ~~`GNUv2`~~ `GNU GPLv2` OSS project and must comply to its own license.
An OSS project cannot statically link in closed source code, which NewTek NDI is.
NewTek NDI provides a dynamic link option which is compatible with OSS, but only for their standard NDI SDK, not for their Advanced SDK.
NDI|HX is only available in the Advanced SDK.

**Reply by dodgepong** [reply]
Small correction, OBS's license is the GNU General Public License v2, aka GPLv2. GNUv2 is not a license.

**Reply by paulpv** [reply]
@dodgepong Thanks for the correction. Updated.

**Reply by MarximusMaximus** [reply]
Ah I see. I didn't realize there wasn't a dynamic link option for NDI-ASDK. That's ... so unusual! I was looking around on the NDI website a bit and noticed that for various cameras, there's a "buy a license" option for HX support for them. I wonder how that's handled.

**Reply by paulpv** [reply]
@MarximusMaximus yup! I suspected that would in the end be the only option that NewTek would approve of for obs-ndi to support HX: hypothetically there would be a "Buy a license" button in the plugin that allowed you to obtain and install a license for the plugin to use.

But even an implementation of that would require a static link in to the closed source NDI-ASDK and thus still violate this plugin's GPLv2 license.

The only option I see around that is to require anyone that wants HX support to have to download the NDI-ASDK themself and then compile the plugin themself.

Basically: We may never be able to provide a prebuilt binary ready to support HX. A user would have to compile it themself and continue to honor the license (meaning their implementation is private and/or can never be redistributed)... but in reality, if someone knows how to do that then they also probably already know the few extra steps required to implement HX in their private fork of the existing code anyway.

**Reply by MarximusMaximus** [reply]
Hm. I'm not totally clear on how GPL2 works out when you want to staticly link against a lib that is closed source. I feel like I've had things do that in the past, but I am probably remembering wrong. I wonder /why/ they don't supply a dynamic lib. To prevent unlicensed linking/usage?

**Reply by dodgepong** [reply]
I have a lot of experience interacting with OBS and the GPLv2, but I am not a lawyer so take all this with a grain of salt.

The issue at hand has more to do with what you're allowed to _distribute_ rather than what you're allowed to _compile_. If you download the SDK and create a build for yourself, there's no issue there. The GPL comes into play the moment you send that build to somebody else, and the GPL grants the recipient certain rights to the source code of what you sent them.

And even then, the issue isn't specifically "static linking", but rather the more ambiguous notion of what constitutes a "derived work". If the two libraries are interacting with each other in a way that there is no meaningful distinction between the two such that they are effectively acting as a single program, then the combined work can only be distributed under the GPL if the licenses of both works are compatible with the GPL. If the two interact in a way that is generally "separate", e.g. through an API designed to be consumed by outside parties, or by invoking `exec` on an executable with CLI flags (like manually running ffmpeg.exe), then the two are generally regarded as being separate works.

Obviously there's a lot of gray area in between these two sides, but it generally comes down to how intimately the two works know about each other's internals, share internal data structures, etc. Static linkage usually represents an intimate combination such that the two works _can't_ be separated. With dynamic linking, IPC, etc. there's less clarity because the distinction becomes semantic, and ultimately requires a judge to decide for sure.

But again, this only "turns on" once you _distribute a build_ of the code. FFmpeg used to support NDI, but only in a way that would be statically linked with the NDI SDK. This required anyone who wanted NDI support in FFmpeg to download the SDK and build it themselves, and this was perfectly fine. However, NewTek decided to start distributing pre-built binaries of FFmpeg with this functionality enabled, and they were rightfully admonished for doing so, to the point of NDI support being _removed_ from FFmpeg and being blackballed from the project. (https://trac.ffmpeg.org/ticket/7589)

So in that regard, I would not feel super confident in reaching out to NewTek about getting special permission for GPL things because they have demonstrated both a lack of understanding of the GPL and a willingness to break the GPL to pursue more users. NewTek's motivation is to _get you using NDI_, and they are not especially motivated to notice that your own license limits the ways you are able to use their library.

**Reply by MarximusMaximus** [reply]
Brainstorming here, maybe not possible regarding the licenses.

Would it be possible to create a BSD3/MIT licensed dynamic library that statically links NDI-ASDK and dynamically soft link (graceful fail if missing) that library from the GPL2 licensed code? And the library would need to be a separate download (like how ffmpeg is a separate download from audacity). 

The dynamic library would be a basic facade to NDI-ASDK, however, note, APIs are (currently) copyrightable in the USA, so if the API closely resembles NDI's, it would need to 100% abide by NDI's copyright license on their ASDK. I am not a lawyer, but, I am not currently seeing anything that would bar this in the license (https://downloads.ndi.tv/SDK/NDI_SDK/NDI%20License%20Agreement.pdf), but admittedly, this license is harder for me to read than some others.

IIRC, nothing prevents GPL2 code from dynamically linking BSD3/MIT code. Unsure how NDI would like a dynamic lib wrapping their static lib though (since, they like, could just have a dynamic lib themselves)... (seriously, why don't they?? is it really just to prevent something else linking against it w/o approval?)

So given that, I think that methodology, regarding NDI's potential requirements:
1) Would there be a way to validate that only the plugin is linking to it and nothing else is (something like verify the name of the linking executable and a hash of the binary)? 
2) Would there be a way to validate (if required by NDI) that someone acquired a license from NDI?

**Reply by jbcurler2010** [reply]
Marx, would you be willing to email me on this topic? jbcurler@gmail.com

**Comment by jbcurler2010** [comment]
Hmmmm. That's troubling.

I would think, given the distro base for OBS and NDI, that NewTek would be
interested in selling all those licenses.

**Comment by jbcurler2010** [comment]
Not sure if anyone here was on the webinar yesterday, but I was able to ask about this. The presenter made a mistake and said that the General SDK would license encoding HX. He corrected himself later. I am following up with them to ask: Since HX3 is now released, would they consider allowing HX2 to run under GPL? Seems reasonable to me.

**Comment by Trouffman** [comment]
Closing this and consolidating in #770

**Reply by s-ol** [reply]
Actually it seems like maybe this one should be the "consolidated issue", given that it has the actual argument and detailed discussion,, whereas #770 just says "look at #825"?

**Comment by s-ol** [comment]
The manual of the NDI Advanced SDK version 6.2 has a section titled "7. DYNAMIC LOADING OF NDI LIBRARIES" which starts with this paragraph:

> At times you might prefer not to link directly to the NDI libraries, loading them dynamically at run time instead. This
can be of value in open source projects. 

~~I haven't checked the unpacked windows SDK yet~~, but the Linux download contains dynamic libraries `libndi_advanced.so.6.2.1`. **EDIT:** yes, the windows SDK comes with a dll file as well.

The section "3. LICENSING" states:

> - You may use the NDI library within free or commercial Products (as defined by License) created using this SDK
without paying any license fees.
> - You should include the NDI DLLs as part of your own application and keep them in your application folders so
that there is no chance that NDI DLLs installed by your application might conflict with other applications on the
system that also use NDI. Please do not install your NDI DLLs into the system path for this reason. If you are
distributing the NDI DLLs you need to ensure that your application complies with the License Agreement, this
section and the license terms outlined in “3rd party rights” towards the end of this manual.
> - The NDI Advanced SDK is provided to allow anyone to develop products against NDI and to develop without any
charge. To use the Advanced SDK in a commercial product or environment you should contact
licensing@ndi.video to receive a vendor ID for your company.

Based on this, it would seem that there is no longer a real blocker for including NDI HX support in DistroAV and other open-source software: it can link dynamically to the NDI Advanced SDK and instruct users to obtain the dynamic libraries themselves.

**Reply by BitRate27** [reply]
Looking into how this would actually work, if this is indeed something NDI will allow, I tried to include the Advanced SDK and load the .dll. This worked fine, just needed to:

- add all the advanced .h files to our repo
- include the Processing.NDI.Advanced.h in plugin-main.h
- put the Processing.NDI.Lib.Advanced.x64.dll into the NDI Runtime folder
- load the Processing.NDI.Lib.Advanced.x64.dll when loading the DistroAV OBS module

I was able to run OBS, load DistroAV into OBS and it ran like normal. 
This does not automatically make the output from DistroAV use HX. To do this, we need to call the send_create_v2 API with a json config parameter as follows:
```
    // JSON configuration
    const char* ndi_config_json = R"json(
    {
      "ndi": {
        "codec": {
          "shq": {
            "mode": "4:2:2",
            "quality": 100
          }
        }
      }
    }
    )json";

    // Create sender
    NDIlib_send_instance_t sender =
        NDIlib_send_create_v2(&send_desc, ndi_config_json);
```
..and this is where we run into issues. The send_create_v2 is not exported in the Advanced SDK for manual dynamic loading, which is how we load standard NDI runtime dll. We don't use the dynamic linker of the OS. Also, the PROCESSINGNDILIB_API is not defined in the Advanced SDK, so we don't have the information to load it like we do the standard .dll.

After some tweaks to get around the missing dynamic links, it crashed when sending audio. So there are some more changes to the actual sending calls to make this work.

I will also note that in the email from NDI when I requested the Advanced SDK, it said that it will only send video for 30 minutes in the trial version of the Advanced SDK.

Will stop here until we get clearance from NDI if this is possible. Just wanted to capture my findings.
